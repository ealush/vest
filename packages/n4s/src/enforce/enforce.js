import compounds from 'compounds';
import { RUN_RULE } from 'enforceKeywords';
import runner from 'enforceRunner';
import isFunction from 'isFunction';
import isRule from 'isRule';
import proxySupported from 'proxySupported';
import rules from 'rules';

const rulesObject = Object.assign(rules(), compounds);

let rulesList = proxySupported() ? null : Object.keys(rulesObject);

const Enforce = value => {
  const proxy = genRuleProxy(enforce, ruleName => (...args) => {
    runner(rulesObject[ruleName], value, args);
    return proxy;
  });
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
    genRuleProxy(Enforce, bindLazyRule);
  }

  return enforce;
};

export default enforce;

// Creates a proxy object that has access to all the rules
function genRuleProxy(target, output) {
  if (proxySupported()) {
    return new Proxy(target, {
      get: (target, fnName) => {
        if (isRule(rulesObject, fnName)) {
          return output(fnName);
        }

        return target[fnName];
      },
    });
  } else {
    /**
     * This method is REALLY not recommended as it is slow and iterates over
     * all the rules for each direct enforce reference. We only use it as a
     * lightweight alternative for the much faster proxy interface
     */
    return rulesList.reduce((target, fnName) => {
      return Object.defineProperties(target, {
        [fnName]: { get: () => output(fnName), configurable: true },
      });
    }, target);
  }
}

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => rulesObject[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
