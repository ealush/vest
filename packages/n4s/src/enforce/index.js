import { isRule, proxySupported } from '../lib';
import rules from '../rules';
import runner from './runner';

const rulesObject = { ...rules };

let enforce, rulesList;

if (proxySupported()) {
  enforce = value => {
    const proxy = new Proxy(rulesObject, {
      get: (rules, fnName) => {
        if (!isRule(rules, fnName)) {
          return;
        }

        return (...args) => {
          runner(rules[fnName], value, ...args);
          return proxy;
        };
      },
    });
    return proxy;
  };
} else {
  rulesList = Object.keys(rulesObject);

  enforce = value =>
    rulesList.reduce(
      (allRules, fnName) =>
        Object.assign(allRules, {
          ...(isRule(rulesObject, fnName) && {
            [fnName]: (...args) => {
              runner(rulesObject[fnName], value, ...args);
              return allRules;
            },
          }),
        }),
      {}
    );
}

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
  }
};

export default enforce;
