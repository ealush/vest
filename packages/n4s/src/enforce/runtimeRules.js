import compounds from 'compounds';
import rules from 'rules';

export default Object.assign(rules(), compounds);
