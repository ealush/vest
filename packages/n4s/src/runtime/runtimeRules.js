import compounds from 'compounds';
import ruleMeta from 'ruleMeta';
import rules from 'rules';

export default Object.assign(rules(), compounds, ruleMeta);
