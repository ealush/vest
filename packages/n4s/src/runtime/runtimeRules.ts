import type { DropFirst } from 'utilityTypes';

import type { TRuleReturn } from 'ruleReturn';
import rules from 'rules';

export type TArgs = any[];

export type TRuleValue = any;

export type TRuleBase = (value: TRuleValue, ...args: TArgs) => TRuleReturn;

export type TRule = Record<string, TRuleBase>;

type TBaseRules = typeof baseRules;
export type KBaseRules = keyof TBaseRules;

const baseRules = rules();

function getRule(ruleName: string): TRuleBase {
  return baseRules[ruleName as KBaseRules];
}

export { baseRules, getRule };

export type TRules<E = Record<string, unknown>> = n4s.EnforceCustomMatchers<
  TRules<E> & E
> &
  Record<string, (...args: TArgs) => TRules<E> & E> & {
    [P in KBaseRules]: (
      ...args: DropFirst<Parameters<TBaseRules[P]>> | TArgs
    ) => TRules<E> & E;
  };

/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/no-empty-interface */
declare global {
  namespace n4s {
    interface IRules<E> extends TRules<E> {}
  }
}
