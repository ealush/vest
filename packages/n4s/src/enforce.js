import bindLazyRule from 'bindLazyRule';
import bindExtend from 'enforce.extend';
import bindTemplate from 'enforce.template';
import runner from 'enforceRunner';
import genRuleProxy from 'genRuleProxy';
import runtimeRules from 'runtimeRules';
import withArgs from 'withArgs';

const Enforce = value => {
  const proxy = genRuleProxy({}, ruleName =>
    withArgs(args => {
      runner(runtimeRules[ruleName], value, args);
      return proxy;
    })
  );
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

bindExtend(enforce, Enforce);
bindTemplate(enforce);

export default enforce;
