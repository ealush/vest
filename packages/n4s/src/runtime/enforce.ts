import assign from 'assign';

import eachEnforceRule from 'eachEnforceRule';
import { ctx, TEnforceContext } from 'enforceContext';
import enforceEager, { TEnforceEager } from 'enforceEager';
import genEnforceLazy, { TLazyRules } from 'genEnforceLazy';
import isProxySupported from 'isProxySupported';
import { TRule, KBaseRules, baseRules, getRule } from 'runtimeRules';
/**
 * Enforce is quite complicated, I want to explain it in detail.
 * It is dynamic in nature, so a lot of proxy objects are involved.
 *
 * Enforce has two main interfaces
 * 1. eager
 * 2. lazy
 *
 * The eager interface is the most commonly used, and the easier to understand.
 * It throws an error when a rule is not satisfied.
 * The eager interface is declared in enforceEager.ts and it is quite simple to understand.
 * enforce is called with a value, and the return value is a proxy object that points back to all the rules.
 * When a rule is called, the value is mapped as its first argument, and if the rule passes, the same
 * proxy object is returned. Otherwise, an error is thrown.
 *
 * The lazy interface works quite differently. It is declared in genEnforceLazy.ts.
 * Rather than calling enforce directly, the lazy interface has all the rules as "methods" (only by proxy).
 * Calling the first function in the chain will initialize an array of calls. It stores the different rule calls
 * and the parameters passed to them. None of the rules are called yet.
 * The rules are only invoked in sequence once either of these chained functions are called:
 * 1. test(value)
 * 2. run(value)
 *
 * Calling run or test will call all the rules in sequence, with the difference that test will only return a boolean value,
 * while run will return an object with the validation result and an optional message created by the rule.
 */

function genEnforce(): TEnforce {
  const target = {
    context: () => ctx.useX(),
    extend: (customRules: TRule) => {
      assign(baseRules, customRules);
    },
  } as TEnforce;

  if (!isProxySupported()) {
    eachEnforceRule((ruleName: KBaseRules) => {
      // Only on the first rule access - start the chain of calls
      target[ruleName] = genEnforceLazy(ruleName);
    });

    return assign(enforceEager, target);
  }

  return new Proxy(assign(enforceEager, target) as TEnforce, {
    get: (target: TEnforce, key: string) => {
      if (key in target) {
        return target[key];
      }

      if (!getRule(key)) {
        return;
      }

      // Only on the first rule access - start the chain of calls
      return genEnforceLazy(key);
    },
  });
}

export const enforce = genEnforce();

type TEnforce = TEnforceMethods & TLazyRules & TEnforceEager;

type TEnforceMethods = {
  context: () => TEnforceContext;
  extend: (customRules: TRule) => void;
};
