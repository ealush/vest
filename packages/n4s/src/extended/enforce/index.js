import enforce from '../../enforce';
import rules, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
} from '../rules';

enforce.extend(rules);

export default Object.assign(enforce, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
});
