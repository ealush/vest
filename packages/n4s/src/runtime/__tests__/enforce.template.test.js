import enforce from 'enforce';
import runtimeRules from 'runtimeRules';

describe('enforce.template', () => {
  test('enforce has `template` static function', () => {
    expect(typeof enforce.template).toBe('function');
  });

  describe('Return value', () => {
    it('Should return a function', () => {
      const x = enforce.template(enforce.isArray(), enforce.longerThan(5));

      expect(typeof x).toBe('function');
    });

    it('Should have a static function called `run`', () => {
      const x = enforce.template(enforce.isArray(), enforce.longerThan(5));

      expect(typeof x.run).toBe('function');
    });
  });

  describe('When the returned function gets called', () => {
    it('Should throw an error when invalid', () => {
      const x = enforce.template(enforce.isArray(), enforce.longerThan(2));
      const y = enforce.template(enforce.isString(), enforce.isNumeric());
      expect(() => x([])).toThrow();
      expect(() => y('')).toThrow();
      expect(() => x([1])).toThrow();
      expect(() => x('hello')).toThrow();
      expect(() => y('hello')).toThrow();
    });

    it('Should return silently when valid', () => {
      const x = enforce.template(enforce.isArray(), enforce.longerThan(2));
      const y = enforce.template(enforce.isString(), enforce.isNumeric());

      x([1, 2, 3]);
      y('123');
    });

    it('Should return rules proxy', () => {
      const x = enforce.template(enforce.isArray());
      expect(x([])).toEqual(enforce(1));
      Object.keys(runtimeRules).forEach(k => {
        expect(typeof x([])[k]).toBe('function');
      });
    });
  });

  describe('When the `run` function gets called', () => {
    it('Should return `false` when invalid', () => {
      const x = enforce.template(enforce.isNumber().greaterThan(50));
      const y = enforce.template(
        enforce.isString().isNumeric(),
        enforce.isOdd()
      );

      expect(x).not.toPassWith(10);
      expect(x).not.toPassWith('90');
      expect(y).not.toPassWith(true);
      expect(x).not.toPassWith('90');
    });
    it('Should return `true` when valid', () => {
      const x = enforce.template(enforce.isNumber().greaterThan(50));
      const y = enforce.template(
        enforce.isString().isNumeric(),
        enforce.isOdd()
      );

      expect(x).toPassWith(51);
      expect(y).toPassWith('11');
    });
  });

  describe('joined templates', () => {
    it('Should validate all joined templates', () => {
      const X = enforce.template(enforce.isString(), enforce.longerThan(2));
      const Y = enforce.template(enforce.isNumeric());
      const Z = enforce.template(enforce.isEven());

      expect(() => enforce.template(X, Y, Z)('a')).toThrow();
      expect(() => enforce.template(X, Y, Z)('abc')).toThrow();
      expect(() => enforce.template(X, Y, Z)('111')).toThrow();
      expect(() => enforce.template(X, Y, Z)(112)).toThrow();
      enforce.template(X, Y, Z)('112');
      expect(enforce.template(X, Y, Z).run('112')).toMatchSnapshot();
      enforce.template(X, Y)('111');
      expect(enforce.template(X, Y).run('111')).toMatchSnapshot();
      enforce.template(X)('abc');
      expect(enforce.template(X).run('abc')).toMatchSnapshot();
    });
  });
});

describe('With nested values', () => {
  const Person = enforce.template(
    enforce.loose({
      name: enforce.loose({
        first: enforce.isString().isNotEmpty(),
        last: enforce.isString().longerThan(2),
        middle: enforce.optional(enforce.isString().isNotEmpty()),
      }),
    })
  );

  const Entity = enforce.template(
    enforce.loose({
      id: enforce
        .isString()
        .startsWith('o.')
        .message('Entity id must be valid'),
    })
  );

  const User = enforce.template(
    enforce.loose({
      name: enforce.loose({ username: enforce.isString().longerThan(3) }),
      friends: enforce
        .isArrayOf(Entity)
        .message('friend id must be a valid entity id'),
    }),
    Entity,
    Person
  );

  it('Should correctly validate template', () => {
    const res_1 = User.run({
      name: {
        first: 'Klaus',
        last: 'Hargreeves',
        username: 'No.5',
      },
      id: 'o.5',
      friends: [
        { id: 'o.1' },
        { id: 'o.2' },
        { id: 'o.3' },
        { id: 'o.4' },
        { id: 'o.6' },
        { id: 'o.7' },
      ],
    });

    expect(res_1.pass).toBe(true);
    expect(res_1).toMatchSnapshot();

    const res_2 = User.run({
      name: {
        first: 'Klaus',
        last: 'Hargreeves',
        username: 'No.5',
      },
      id: 'o.5',
      friends: [{ id: 'o.1' }, { id: 'invalid_id' }],
    });

    expect(res_2.pass).toBe(false);
    expect(res_2).toMatchSnapshot();

    const res_3 = User.run({
      name: {
        first: 'Klaus',
        last: 'Hargreeves',
        username: '#5',
      },
      id: 'x5',
      friends: [{ id: 'o.1' }],
    });

    expect(res_3.pass).toBe(false);
    expect(res_3).toMatchSnapshot();
  });
});
