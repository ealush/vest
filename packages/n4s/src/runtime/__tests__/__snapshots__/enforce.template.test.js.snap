// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With nested values Should correctly validate template 1`] = `
RuleResult {
  "children": Object {
    "friends": RuleResult {
      "children": Object {
        "0": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "1": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "2": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "3": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "4": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "5": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "isArray": true,
      "message": "friend id must be a valid entity id",
    },
    "id": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "message": "Entity id must be valid",
      "warn": false,
    },
    "name": RuleResult {
      "children": Object {
        "first": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "last": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "middle": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "username": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
    },
  },
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
}
`;

exports[`With nested values Should correctly validate template 2`] = `
RuleResult {
  "children": Object {
    "friends": RuleResult {
      "children": Object {
        "0": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
        "1": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": true,
              "hasErrors": true,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
      "isArray": true,
      "message": "friend id must be a valid entity id",
    },
    "id": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "message": "Entity id must be valid",
      "warn": false,
    },
    "name": RuleResult {
      "children": Object {
        "first": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "last": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "middle": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "username": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`With nested values Should correctly validate template 3`] = `
RuleResult {
  "children": Object {
    "friends": RuleResult {
      "children": Object {
        "0": RuleResult {
          "children": Object {
            "id": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "Entity id must be valid",
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "isArray": true,
      "message": "friend id must be a valid entity id",
    },
    "id": RuleResult {
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
      "message": "Entity id must be valid",
      "warn": false,
    },
    "name": RuleResult {
      "children": Object {
        "first": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "last": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "middle": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "username": RuleResult {
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
          "warn": false,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`enforce.template joined templates Should validate all joined templates 1`] = `
RuleResult {
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
  "warn": false,
}
`;

exports[`enforce.template joined templates Should validate all joined templates 2`] = `
RuleResult {
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
  "warn": false,
}
`;

exports[`enforce.template joined templates Should validate all joined templates 3`] = `
RuleResult {
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
  "warn": false,
}
`;
